### ISSUE TITLE
Hallucination Prevention & Detection: retrieval gate, system prompts, citation check hook

### ISSUE THREAD

**Issue by AndreLiar** [andrelair-platform/platform-backlog#44]
## Context

The evaluation framework in #43 measures hallucination after the fact (Ragas faithfulness nightly, LLM-as-judge). This issue adds three real-time layers that operate during the request lifecycle:

1. **Retrieval grounding gate** — refuse the call before it reaches Ollama if no relevant documents were found (prevents free hallucination on empty context)
2. **Department system prompts** — constrain the model at the prompt level to cite sources and say "I don't know" rather than invent
3. **Citation check hook** — post-generation flag for high-stakes departments when the model makes a specific claim (amount, article, percentage) without citing a source

**Depends on:** #34 (LiteLLM gateway), #38 (Qdrant + retrieval pipeline), #37 (LiteLLM hook framework)

---

## Insurance Hallucination Risk Classification

Before implementation — map risk by query type so controls are proportional:

| Query type | Hallucination risk | Why dangerous |
|---|---|---|
| Claim amount / indemnity limit | **CRITICAL** | Wrong amount paid to claimant — legal liability |
| Policy clause / article reference | **CRITICAL** | Non-existent clause cited — contract dispute |
| French regulation (Code des assurances) | **HIGH** | Invented article numbers — compliance failure |
| Premium calculation | **HIGH** | Wrong price quoted — financial loss |
| Reinsurance treaty terms | **HIGH** | Treaty doesn't exist or wrong retention |
| Coverage territory / exclusions | **HIGH** | Customer told they're covered when they aren't |
| IT/DevOps question | **LOW** | Wrong command — easily verified |
| General HR / admin | **LOW** | Inconvenience only |

High-stakes departments: `direction-sinistres`, `direction-finance`, `direction-juridique`, `direction-souscription`, `direction-actuariat`, `direction-audit`, `direction-réassurance`.

---

### Problem

When Qdrant returns no results above the similarity threshold, LiteLLM still forwards the request to Ollama. The model then answers from training data alone — confidently hallucinating policy terms, claim procedures, or regulatory references that sound plausible but are fabricated.

### Implementation: `retrieval_grounding_hook.py`

```python
import re
from litellm.integrations.custom_logger import CustomLogger
from fastapi import HTTPException

SIMILARITY_THRESHOLD = 0.35      # below this = "no relevant documents found"
CITATION_REQUIRED_DEPTS = {
    "direction-sinistres",
    "direction-finance",
    "direction-juridique",
    "direction-souscription",
    "direction-actuariat",
    "direction-audit",
    "direction-réassurance",
}

class RetrievalGroundingHook(CustomLogger):
    """
    Pre-call: refuse requests where retrieval returned no relevant context.
    Post-call: flag high-stakes responses that make specific claims without citing.
    """

    async def async_pre_call_hook(self, user_api_key_dict, cache, data, call_type):
        metadata = data.get("metadata", {})
        dept = user_api_key_dict.metadata.get("department", "")

        # Qdrant retrieval score passed through Open WebUI metadata
        # (set by the RAG pipeline in Open WebUI before sending to LiteLLM)
        max_similarity = metadata.get("qdrant_max_similarity")

        if max_similarity is not None and max_similarity < SIMILARITY_THRESHOLD:
            dept_label = dept.replace("direction-", "").title()
            raise HTTPException(
                status_code=400,
                detail=(
                    f"Aucun document pertinent trouvé pour votre question "
                    f"(score de similarité {max_similarity:.2f} < {SIMILARITY_THRESHOLD}). "
                    f"Veuillez reformuler ou contacter votre référent {dept_label}."
                ),
            )

    async def async_post_call_success_hook(self, data, user_api_key_dict, response):
        dept = user_api_key_dict.metadata.get("department", "")
        if dept not in CITATION_REQUIRED_DEPTS:
            return

        text = response.choices[0].message.content or ""
        trace_id = response._hidden_params.get("litellm_call_id", "")

        # Detect specific factual claims
        CLAIM_PATTERNS = [
            r"\d[\d\s]*[.,]\d{2}\s*€",            # monetary amounts: 12 500,00 €
            r"\d+\s*%",                             # percentages
            r"article\s+[lra]\s*[\d\-]+",          # French law articles: Article L111-1
            r"clause\s+\d+",                        # contract clauses
            r"(délai|deadline)\s+(de\s+)?\d+\s*(jour|mois|an)", # time limits
            r"(couvert|garanti|exclu)\s+(jusqu|à hauteur|pour)",  # coverage assertions
        ]
        CITATION_MARKERS = [
            "selon", "d'après", "source", "document", "article",
            "clause", "contrat", "police", "avenant", "annexe",
            "référence", "voir", "cf.", "conformément",
        ]

        has_specific_claim = any(
            re.search(p, text, re.IGNORECASE) for p in CLAIM_PATTERNS
        )
        has_citation = any(m in text.lower() for m in CITATION_MARKERS)

        if has_specific_claim and not has_citation:
            # Non-blocking — flag for review, don't reject the response
            from langfuse import Langfuse
            import os
            lf = Langfuse(
                public_key=os.environ["LANGFUSE_PUBLIC_KEY"],
                secret_key=os.environ["LANGFUSE_SECRET_KEY"],
                host=os.environ.get("LANGFUSE_HOST", "http://langfuse.ai.svc:3000"),
            )
            lf.event(
                trace_id=trace_id,
                name="uncited-specific-claim",
                level="WARNING",
                metadata={
                    "department": dept,
                    "response_excerpt": text[:300],
                    "action": "flagged-for-review",
                },
            )
            lf.score(
                trace_id=trace_id,
                name="citation_compliance",
                value=0,
                comment="Specific claim detected without source citation",
                data_type="NUMERIC",
            )
        elif has_specific_claim and has_citation:
            from langfuse import Langfuse
            import os
            lf = Langfuse(
                public_key=os.environ["LANGFUSE_PUBLIC_KEY"],
                secret_key=os.environ["LANGFUSE_SECRET_KEY"],
                host=os.environ.get("LANGFUSE_HOST", "http://langfuse.ai.svc:3000"),
            )
            lf.score(
                trace_id=trace_id,
                name="citation_compliance",
                value=1,
                comment="Specific claim cited correctly",
                data_type="NUMERIC",
            )
```

### Register in LiteLLM ConfigMap

```yaml
# In litellm-config ConfigMap (from issue #34)
litellm_settings:
  success_callback: ["langfuse", "otel"]
  failure_callback: ["langfuse"]
  callbacks:
    - /hooks/data_sovereignty_hook.RetrievalGroundingHook  # add alongside existing hooks
```

---

### Problem

Default Ollama system prompts have no grounding constraint. `llama3.2:3b` will confidently answer any question using training data, including inventing French insurance regulations (Code des assurances, Code civil) that sound authoritative.

### Implementation: system prompt per department in LiteLLM router

Add to the LiteLLM router config (extends issue #34 ConfigMap):

```yaml
model_list:
  - model_name: claims-assistant
    litellm_params:
      model: ollama/llama3.2:3b
      api_base: http://ollama-set-hog.ai.svc:11434
      system_prompt: |
        Tu es un assistant spécialisé pour la direction sinistres de HDI Seguros France.
        
        RÈGLES ABSOLUES — tu ne peux jamais les enfreindre:
        1. Réponds UNIQUEMENT à partir des documents fournis dans le contexte ci-dessus.
        2. Si le contexte ne contient pas l'information demandée, réponds exactement:
           "Cette information ne figure pas dans mes documents. Merci de contacter votre référent sinistres."
        3. N'invente JAMAIS: montants d'indemnisation, délais réglementaires, références d'articles,
           clauses contractuelles, noms de garanties.
        4. Pour toute valeur chiffrée (montant, délai, pourcentage), indique toujours le document source.
        5. En cas de doute sur l'interprétation d'une clause, dis-le explicitement.

  - model_name: finance-assistant
    litellm_params:
      model: ollama/llama3.2:3b
      api_base: http://ollama-set-hog.ai.svc:11434
      system_prompt: |
        Tu es un assistant pour la direction finance et comptabilité de HDI Seguros France.
        
        RÈGLES ABSOLUES:
        1. Réponds uniquement à partir du contexte documentaire fourni.
        2. Toute donnée chiffrée (montant, ratio, provision, taux) doit être accompagnée de sa source.
        3. Ne jamais déduire ou extrapoler des montants financiers non explicitement présents dans le contexte.
        4. Si la question porte sur des normes IFRS17, Solvabilité II ou des règles fiscales,
           cite l'article ou la norme exacte — sinon indique que tu n'as pas cette information.
        5. Les projections ou estimations ne sont jamais de ta responsabilité.

  - model_name: legal-assistant
    litellm_params:
      model: ollama/llama3.2:3b
      api_base: http://ollama-set-hog.ai.svc:11434
      system_prompt: |
        Tu es un assistant pour la direction juridique de HDI Seguros France.
        
        RÈGLES ABSOLUES:
        1. Tu n'es PAS un conseil juridique — précise-le si on te demande une opinion de droit.
        2. Cite uniquement les textes explicitement présents dans les documents fournis.
        3. Ne jamais paraphraser ou résumer un article de loi sans l'avoir dans ton contexte.
        4. Si tu ne trouves pas la réponse dans les documents: "Cette question requiert une analyse
           juridique par un membre de la direction juridique."
        5. Signale toujours si la législation a pu évoluer depuis la date du document source.

  - model_name: general-assistant
    litellm_params:
      model: ollama/llama3.2:1b
      api_base: http://ollama-set-hog.ai.svc:11434
      system_prompt: |
        Tu es un assistant pour les équipes de HDI Seguros France.
        Réponds de manière concise et professionnelle.
        Si tu n'es pas sûr d'une information, dis-le clairement plutôt que de deviner.
```

### Virtual key → model routing (extends issue #34)

```yaml
# Map each department's virtual key to its dedicated model_name
# In LiteLLM admin config:
router_settings:
  routing_strategy: usage-based-routing
  model_group_alias:
    direction-sinistres:   claims-assistant
    direction-finance:     finance-assistant
    direction-juridique:   legal-assistant
    direction-souscription: claims-assistant   # shares claims model
    direction-actuariat:   finance-assistant   # shares finance model
    direction-audit:       finance-assistant
    direction-réassurance: finance-assistant
    direction-it:          general-assistant
    direction-rh:          general-assistant
    direction-commercial:  general-assistant
```

---

### Problem

`llama3.2:3b` rarely says "I don't know" — it prefers to generate a plausible-sounding answer even at low confidence. The system prompt helps but doesn't fully solve this.

### Implementation: post-call confidence signal

Add a lightweight second Ollama call to rate the model's own confidence when serving high-stakes departments. Only runs if the response contains CLAIM_PATTERNS (reuses the detection logic from Layer 1):

```python
SELF_EVAL_PROMPT = """Evaluate your previous response on a scale:
1 = I invented this, not grounded in any document
2 = I'm uncertain but made a reasonable inference
3 = I'm confident this is directly supported by the provided context

Respond with JSON only: {{"confidence": <1|2|3>, "reason": "<10 words max>"}}"""

async def self_evaluate(question: str, answer: str, context: str) -> dict:
    result = completion(
        model="ollama/llama3.2:1b",   # use fast model for meta-eval
        api_base="http://ollama-set-hog.ai.svc:11434",
        messages=[
            {"role": "user", "content": question},
            {"role": "assistant", "content": answer},
            {"role": "user", "content": SELF_EVAL_PROMPT},
        ],
        response_format={"type": "json_object"},
        max_tokens=60,
    )
    import json
    return json.loads(result.choices[0].message.content)
```

If confidence == 1 → prepend warning to response:
> ⚠️ *Je ne dispose pas de documents suffisants pour répondre avec certitude à cette question. Veuillez vérifier auprès de votre référent métier.*

---

## Grafana Dashboard Panel — Hallucination Risk Monitor

Add to the AI Observability dashboard (extends monitoring from #36 and #43):

```
Panel: Hallucination Risk by Department (last 24h)
────────────────────────────────────────────────────
Metric 1: ragas_faithfulness per department (heatmap)
Metric 2: uncited_specific_claim_total per department (bar)
Metric 3: retrieval_gate_refused_total per department (line)
Metric 4: citation_compliance score average per department (gauge)

Alert: if uncited_specific_claim for finance or juridique > 5 in 1h → PagerDuty
```

---

## Acceptance Criteria

- [ ] Send a query to Claims dept with no matching documents in Qdrant → LiteLLM returns 400 with French "no documents found" message, trace shows `retrieval_gate_blocked=true` in Langfuse
- [ ] Send a valid query to Finance dept → response contains a euro amount → Langfuse shows `citation_compliance=0` WARNING event if no source cited, `citation_compliance=1` if cited
- [ ] System prompt for `claims-assistant` model visible in LiteLLM admin UI under model configuration
- [ ] Ask `direction-it` a general coding question → no citation check runs (dept not in CITATION_REQUIRED_DEPTS)
- [ ] Ragas faithfulness < 0.70 on a Claims trace after retrieval gate → alert fires within 20 min (fast CronJob from #43)
- [ ] Self-evaluation confidence=1 response → prepended warning visible in Open WebUI response bubble
- [ ] Grafana panel shows per-dept `uncited-specific-claim` event count in last 24h
- [ ] Langfuse `citation_compliance` score visible as a column in the Traces table for sinistres/finance/juridique
- [ ] All 3 hooks registered in LiteLLM ConfigMap alongside existing `DataSovereigntyHook` from #42

---

## What This Does NOT Cover

- **Factual accuracy against external ground truth** — for that, you need a curated golden dataset per department (issue #43, Langfuse Datasets). The citation check only verifies *whether a source was cited*, not *whether the source is correct*.
- **Temporal hallucination** (model uses outdated training data for current regulations) — mitigated by the system prompt instruction to flag document dates, not eliminated.
- **Cross-document contradiction** (two retrieved docs disagree) — deferred; requires a multi-document consistency check that Ragas doesn't currently cover.

**Est:** 3–4 hr
**Depends on:** #34 (LiteLLM), #37 (hook framework), #38 (Qdrant similarity scores passed through metadata), #43 (Langfuse scoring infrastructure)
